ensime_shared/ensime.py

```diff
@@ -584,9 +584,13 @@ def handle_completion_info_list(self, call_id, payload):
         """Handler for a completion response."""
         completions = payload["completions"]
         self.log("handle_completion_info_list: in")
-        self.suggestions = [completion_to_suggest(c) for c in completions]
+        self.suggestions = [completion_to_suggest(c, self.format_type_params()) for c in completions]
         self.log("handle_completion_info_list: {}".format(self.suggestions))
 
+    def format_type_params(self):
+        """NOP as it's not available for Server 1.0"""
+        return lambda _: ""
+
```

Comment:
It is a little bit magical, but not metaclass hackery exactly:

```python
>>> class ProtocolHandler(object):
...     pass
...
>>> class ProtocolHandlerV1(ProtocolHandler):
...     pass
...
>>> class ProtocolHandlerV2(ProtocolHandler):
...     pass
...
>>> class OtherMixins(object):
...     pass
...
>>> class EnsimeClient(OtherMixins, object):
...     pass
...
>>> server_v2 = True
>>> def during_startup():
...     if server_v2 == True:
...         Protocol = ProtocolHandlerV2
...     else:
...         Protocol = ProtocolHandlerV1
...     EnsimeClient.__bases__ = (Protocol,) + EnsimeClient.__bases__
...
>>> during_startup()
>>> client = EnsimeClient()
>>> client.__class__.__bases__
(<class '__main__.ProtocolHandlerV2'>, <class '__main__.OtherMixins'>, <class 'object'>)
```

Some Pythonistas balk at almost any form of metaprogramming, but this isn't all that uncommon. But yeah, I suppose there's little gain over something without the magic, like:

```python
>>> class EnsimeClientV1(ProtocolHandlerV1, EnsimeClient):
...     pass
...
>>> class EnsimeClientV2(ProtocolHandlerV2, EnsimeClient):
...     pass
...
>>> def factory():
...     if server_v2 == True:
...         return EnsimeClientV2()
...     else:
...         return EnsimeClientV1()
...
>>> client = factory()
```